### DISCUSSION TITLE
Issue with Unit test SixLabors.ImageSharp.Image.Load

### DISCUSSION THREAD

**Discussion by aakesh001** [SixLabors/ImageSharp#1274] [Q&A]
HI 
I m trying to moq IFormFile object for a unit test in .net core application but the moq always fails on this line SixLabors.ImageSharp.Image.Load(postedFile.OpenReadStream(), out SixLabors.ImageSharp.Formats.IImageFormat format)).

Tried everything Read real image
Created custom Byte 

Always fails below is my moq code
 public void CreateCompany_Success()
        {
        var formFile = new Mock<IFormFile>();
            var sourceImg = File.OpenRead(@"Assets\SampleIcon.png");
            var ms = new MemoryStream();
            var writer = new StreamWriter(ms);
            writer.Write(sourceImg);
            writer.Flush();
            ms.Position = 0;
            var fileName = "logo.png";
            formFile.Setup(f => f.FileName).Returns(fileName).Verifiable();
            formFile.Setup(_ => _.CopyToAsync(It.IsAny<Stream>(), It.IsAny<CancellationToken>()))
                .Returns((Stream stream, CancellationToken token) => ms.CopyToAsync(stream))
                .Verifiable();

            var contentType = "image/png";
            formFile.Setup(_ => _.FileName).Returns(fileName);
            formFile.Setup(_ => _.Length).Returns(ms.Length);
            formFile.Setup(_ => _.OpenReadStream()).Returns(GetBitmapData());
            formFile.Setup(_ => _.ContentType).Returns(contentType);

           _companyService.CheckValidFileDimensionsLinux(formFile);
          _companyProviderMock.Verify(r => r.CreateCompany(It.IsAny<CompanyData>()), Times.Once);
}

public  bool CheckValidFileDimensionsLinux(this IFormFile postedFile, ImageDimension imageDimension = null)
        {
            try
            {
                using (var image = SixLabors.ImageSharp.Image.Load(postedFile.OpenReadStream(), out 
                     SixLabors.ImageSharp.Formats.IImageFormat format))
                {
                    if (imageDimension != null && !ValidDimensionsLinux(image, imageDimension))
                    {
                        return false;
                    }
                }
            }
            catch (Exception ex)
            {
                return false;
            }
            finally
            {
                postedFile.OpenReadStream().Position = 0;
            }
            return true;
        }

**Comment by JimBobSquarePants** [comment]
>  the moq always fails on this line

What's the stack trace? You've posted no meaningful information. Can you please properly format  your code samples also.

I'm curious to why you're even mocking `IFormFile`. It looks like you're having to configure far more boilerplate code than simply creating an instance of `FormFile`.

**Comment by aakesh001** [comment]
I m mocking the Iformfile because when in the Jenkins pipeline we don't want to create any dependency on the actual file.

I have updated the above code sample let me know if you need more information.

I m using real image which is of PNG format
This is the error message

"Image cannot be loaded. Available decoders:\r\n - JPEG : JpegDecoder\r\n - TGA : TgaDecoder\r\n - GIF : GifDecoder\r\n - PNG : PngDecoder\r\n - BMP : BmpDecoder\r\n"

This is the stack trace   

 at SixLabors.ImageSharp.Image.Load(Configuration configuration, Stream stream, IImageFormat& format)
   at SixLabors.ImageSharp.Image.Load(Stream stream, IImageFormat& format)
   at Soti.EMP.Server.Utility.Extensions.FileUploadExtensions.CheckValidFileDimensionsLinux(IFormFile postedFile, ImageDimension imageDimension) in D:\Repositry\SotiCentralGit\MarketPlaceServer\Soti.EMP.Server.Utility\Extensions\FileUploadExtensions.cs:line 119

**Reply by JimBobSquarePants** [reply]
Looks like your input buffer doesn't contain valid image data. I'm assuming your input stream is in the correct position?

Here's some code I've used previously to test against an `IFormFile` parameter

``` c#
using (var ms = new MemoryStream(CreateImageBuffer()))
{
    IFormFile file =  new FormFile(ms, 0, ms.Length, "TestMedia", "TestMedia.gif"),
}

private static byte[] CreateImageBuffer()
{
    // 1x1 pixel black gif
    // data:image/gif;base64,R0lGODlhAQABAIABAEdJRgAAACwAAAAAAQABAAACAkQBAA==
    return Convert.FromBase64String("R0lGODlhAQABAIABAEdJRgAAACwAAAAAAQABAAACAkQBAA==");
}

```

What exactly are you trying to test? `CheckValidFileDimensionsLinux` looks like you're trying to test our code not yours which is not the purpose of a unit test.

This is a mocking issue and not really anything to do with ImageSharp so I'm not sure how I can help you.
